### ISSUE TITLE
BUG : CDN script compatibility with Control Plane (SourceConfig) / versioning

### ISSUE THREAD

**Issue by natenho** [rudderlabs/rudder-sdk-js#1161]
**Describe the bug**
Hi, I know that the v1 is deprecated, but could you please provide the v1 script in the CDN? I have an application that is not dependent on control plane, and in some version the events stopped to be sent because of writeKey validation.

**To Reproduce**

Using the latest version of SDK, try to track events without depending on control plane (sourceConfig)

**Expected behavior**
If possible, cdn.rudderlabs.com/v1 should return a script compatible with V1 instead of V2
If not possible, another file name could be acceptable as well, like `rudder-analytics-1.33.0.min.js`
It would be nice if all releases were available through CDN so we could select earlier versions.

**Screenshots**
This is the current behavior
![image](https://github.com/rudderlabs/rudder-sdk-js/assets/4236481/af14ef21-e7dc-4657-ab58-c887e5826c49)

Previous versions would send events even if sourceConfig response was any error.

**Additional Information (please complete the following information):**

- SDK installation: CDN
- CDN URL: https://cdn.rudderlabs.com/v1.1/rudder-analytics.min.js

Could you also please share why new versions are depending on the sourceConfig to send events? In which version that was implemented (I could not find myself, so the breaking change could be introuduced in 2.xx as well). Another guess  is that it was a control plane implementation change, not the SDK.

**Comment by natenho** [comment]
@MoumitaM Using git bisect I've figured out that the problem was introduced in v2.23.0 (seems to be #718) so it is caused by this recent change.

Evidences

**Working v2.22.3 **
![image](https://github.com/rudderlabs/rudder-sdk-js/assets/4236481/e8636e12-0fe7-49ce-a921-ef785fa9bf52)

**Not Working v2.23.0 and later**
![image](https://github.com/rudderlabs/rudder-sdk-js/assets/4236481/90653926-dc02-4dc9-8af8-2192e39dd637)

**Comment by MoumitaM** [comment]
@natenho, SDK is always dependent on sourceConfig endpoint. You are passing the wrong writekey. That is why you are getting this error.

> I have an application that is not dependent on control plane

I didn't get this. Do you use the `getSourceConfig` load option?

The change we introduced in this [PR](https://github.com/rudderlabs/rudder-sdk-js/pull/718) is a new feature and not a breaking change.

**Comment by natenho** [comment]
@MoumitaM yes, `getSourceConfig` can illustrate that (Perhaps I should be using it, but I was not).

Using getSourceConfig would "bypass" the dependency on control plane /sourceConfig endpoint, right?
> getSourceConfig | Returns a custom configuration which can be used in place of the control plane’s dashboard

So, consider the following load code inside the [regular rudderstack snippet](https://www.rudderstack.com/docs/sources/event-streams/sdks/rudderstack-javascript-sdk/quick-start-guide/):

```
<script>
// ...
      rudderanalytics.load(
        "my-custom-wk",
        "http://localhost:5000/sdk",
        {
          getSourceConfig: function () { // Do not depend on control plane
            return "this might generate an error but ok, events are sent anyway in v2.22.3"; // intentional!
          },
        }
      );

// Code for identify()
// Code for page()
// Code for track()
</script>
```

Using the version 2.22.3, this is what happens:

![image](https://github.com/rudderlabs/rudder-sdk-js/assets/4236481/8ae301f6-f500-4bc0-b560-11010b7602bc)

Notice the 3 events requests are sent  (identify, page, track), even though there was an error:

![image](https://github.com/rudderlabs/rudder-sdk-js/assets/4236481/f09c972e-f567-4308-a49d-bc7b48bc34ec)

Now, switching to version 2.23.0 (and later), this is what happens:

![image](https://github.com/rudderlabs/rudder-sdk-js/assets/4236481/1fc4d19b-30f0-465c-b353-ef3baa5e51ce)

I'm curious whether the previous behavior was unintentional or if the current behavior is a bug.

**Comment by saikumarrs** [comment]
Hi @natenho 👋 
In [v2.23.0](https://github.com/rudderlabs/rudder-sdk-js/releases/tag/v2.23.0), we've introduced the [data residency feature](https://github.com/rudderlabs/rudder-sdk-js/pull/718) in the SDK, where all the events, including the pre-load buffered events, are processed only after a successful source configuration call. Because the data plane URL is determined only then, it takes the highest precedence over the data plane URL specified in the load API call.

I think your use case is unconventional and generally not recommended. Also, the functionality of the SDK cannot be changed now as it'll be a breaking change for customers using the data residency feature for some edge cases.

However, one workaround is to supply a custom (dummy) source configuration response in the load API options.
https://www.rudderstack.com/docs/sources/event-streams/sdks/rudderstack-javascript-sdk/load-js-sdk/#loading-options

Example:
```
rudderanalytics.load(<writeKey>, <dataplaneUrl>, {
  getSourceConfig: function () {
    // minimal source configuration response to satisfy the SDK
    return {
      source: {
        id: "some-value",
        config: {},
        destinations: []
      }
    };
  }
});
```

**Comment by natenho** [comment]
@saikumarrs, thanks. I noticed that could be an option, but I was trying to avoid that workaround. For now, I'll stick with a self-built and hosted v2.22.3. It would still be good to have earlier versions available through the official CDN, though.

I thought of proposing a way to disable the data residency feature, but future features could also exhibit similar behavior.

As I mentioned earlier, I have an unconventional scenario, so if the above might not be of interest to you (specially the CDN part), feel free to close this issue.

**Comment by bardisg** [comment]
Hello @natenho,

I am late to the conversation, but wanted to flag out that v1 version is not supported but it is not, and will not, be removed from the CDN. You can always retrieve it using https://cdn.rudderlabs.com/v1/rudder-analytics.min.js

Having said that, it is highly advisable to NOT use v1 anymore since it lacks in features, bug & security fixes.

V2.22.3 can also be retrieve from the CDN at https://cdn.rudderlabs.com/2.22.3/rudder-analytics.min.js incase you do not want to self host it.

What @saikumarrs suggested is the best option that will also allow you keep using latest and new releases.

**Comment by natenho** [comment]
Hello @bardisg, thank you for your response!

Your first sentence is closely related to my first concern when I opened the issue. I noticed an inconsistency in the versioning of CDN URLs:

https://cdn.rudderlabs.com/v1/rudder-analytics.min.js - Returns the SDK 1.33.0 ✅
https://cdn.rudderlabs.com/v1.1/rudder-analytics.min.js - Returns the SDK 2.35.1 ❌
![image](https://github.com/rudderlabs/rudder-sdk-js/assets/4236481/22ea16b5-fe05-4530-b6b2-756f0f2773d6)

It is very good to know that there is the https://cdn.rudderlabs.com/2.22.3/rudder-analytics.min.js available!!! (Although it deviates from the previous URL version pattern - I'd expect v2.22.3)
Is that documented anywhere? 

@saikumarrs's suggestion is excellent, but since I have a custom implementation, the approach may not guarantee full compatibility with future versions. I'd rather lock into a specific version (similar to how a distribution with the NPM package would work for clients) so that I can have more control over upgrades.
